gobblin-runtime/src/main/java/gobblin/runtime/JobContext.java

```diff
@@ -170,38 +169,131 @@ public boolean isJobLockEnabled() {
   }
 
   /**
-   * Get an instance of {@link StateStore} for serializing/deserializing {@link JobState}.
-   *
-   * @return an instance of {@link StateStore} for serializing/deserializing {@link JobState}
+   * Store job execution information into the job history store.
    */
-  public StateStore<JobState> getJobStateStore() {
-    return this.jobStateStore;
+  void storeJobExecutionInfo() {
+    if (this.jobHistoryStoreOptional.isPresent()) {
+      try {
+        this.jobHistoryStoreOptional.get().put(this.jobState.toJobExecutionInfo());
+      } catch (IOException ioe) {
+        this.logger.error("Failed to write job execution information to the job history store: " + ioe, ioe);
+      }
+    }
   }
 
   /**
-   * Get an {@link Optional} of {@link JobHistoryStore}.
-   *
-   * @return an {@link Optional} of {@link JobHistoryStore}
+   * Set final {@link JobState} of the given job.
    */
-  public Optional<JobHistoryStore> getJobHistoryStoreOptional() {
-    return this.jobHistoryStoreOptional;
+  void setFinalJobState() {
+    this.jobState.setEndTime(System.currentTimeMillis());
+    this.jobState.setDuration(this.jobState.getEndTime() - this.jobState.getStartTime());
+
+    for (TaskState taskState : this.jobState.getTaskStates()) {
+      // Set fork.branches explicitly here so the rest job flow can pick it up
+      this.jobState.setProp(ConfigurationKeys.FORK_BRANCHES_KEY,
+          taskState.getPropAsInt(ConfigurationKeys.FORK_BRANCHES_KEY, 1));
+    }
+
+    if (this.jobState.getState() == JobState.RunningState.SUCCESSFUL) {
+      // Reset the failure count if the job successfully completed
+      this.jobState.setProp(ConfigurationKeys.JOB_FAILURES_KEY, 0);
+    }
+
+    if (this.jobState.getState() == JobState.RunningState.FAILED) {
+      // Increment the failure count by 1 if the job failed
+      int failures = this.jobState.getPropAsInt(ConfigurationKeys.JOB_FAILURES_KEY, 0) + 1;
+      this.jobState.setProp(ConfigurationKeys.JOB_FAILURES_KEY, failures);
+    }
+  }
+
+  /**
+   * Commit the job on a per-dataset basis.
+   */
+  void commit() throws Exception {
+    Map<String, JobState.DatasetState> datasetStatesByUrns = this.jobState.getDatasetStatesByUrns();
+    for (Map.Entry<String, JobState.DatasetState> entry : datasetStatesByUrns.entrySet()) {
+      setFinalDatasetState(entry.getValue());
+      if (canCommitDataset(this.jobCommitPolicy, entry.getValue())) {
+        try {
+          commitDataset(entry.getValue());
+        } finally {
```

Comment:
I think you should catch the exceptions and log/save them. At the end, you can check if there was an exception thrown and rethrow it. This will allow us to commit as many dataset states even if there was an issue with some of them.